Restore API key and query param overrides

pakrym-oai opened the pull request:
Addresses https://github.com/openai/codex/issues/1796

pakrym-oai committed:
Restore API key and query param overrides

pakrym-oai committed:
core/client: fix auth clone bug by removing unnecessary as_ref() call

pakrym-oai committed:
tests: use custom provider in azure_overrides_assign_properties_used_for_responses_url

Replace "openai" with "custom" as provider name and remove explicit model assignment to improve test realism.

pakrym-oai commented on codex-rs/core/tests/client.rs:
```diff
@@ -376,6 +379,77 @@ async fn includes_user_instructions_message_in_request() {
+    };
+
+    unsafe {
+        std::env::set_var("TEST_API_KEY_ENV_VAR", "Test API Key");
```
Hm, in this case the implementation is supposed to read from the variable. I can remove it from the test but then we'll lose coverage. 

What did you do for other cases where reading of the env variable is required ?

bolinfest replied:
The situation was slightly different, but we supported the values via `ModelProviderInfo` instead: https://github.com/openai/codex/pull/1601

bolinfest replied:
I doesn't appear that you are verifying that the `Authorization: Bearer` header is sent, are you?

pakrym-oai replied:
I am now.

pakrym-oai commented on codex-rs/core/tests/client.rs:
```diff
@@ -376,6 +379,77 @@ async fn includes_user_instructions_message_in_request() {
+    wait_for_event(&codex, |ev| matches!(ev, EventMsg::TaskComplete(_))).await;
+
+    // If URL or query params were wrong, the expectation would fail; also assert single request
+    let requests = server.received_requests().await.unwrap();
```
wait_for_event will fail as well, removed

pakrym-oai committed:
fb

pakrym-oai committed:
am now

pakrym-oai commented on codex-rs/core/tests/client.rs:
```diff
@@ -376,6 +379,77 @@ async fn includes_user_instructions_message_in_request() {
+    let provider = ModelProviderInfo {
+        name: "custom".to_string(),
+        base_url: Some(format!("{}/openai", server.uri())),
+        env_key: Some("TEST_API_KEY_ENV_VAR".to_string()),
```
Then we'll be testing `env_value` because the `env_key` entire codepath will have to be skipped.

bolinfest replied:
Tests are about risk management.

I think adding this `unsafe` block introduces more risk than not having test coverage for `env_key` because if we stopped honoring `env_key`, I suspect things would break pretty quickly and we would resolve the issue in no time. By comparison, the `unsafe` block could cause tests to be flaky, which is much harder to debug.

The other [somewhat silly] thing we can do is choose an environment variable that we expect to exist, like `PATH` and use that as the `env_key`. It appears to have worked thus far:

https://github.com/openai/codex/blob/be41458a498499ebba2f953b73752103ed2c960e/codex-rs/core/tests/stream_no_completed.rs#L80-L83

pakrym-oai replied:
Removed the unsafe code.

pakrym-oai commented on codex-rs/core/src/client.rs:
```diff
@@ -180,30 +181,33 @@ impl ModelClient {
         let max_retries = self.provider.request_max_retries();
 
+        // Precompute query string once for all retry attempts.
+        let query_string = self.provider.get_query_string();
```
I'd rather move it to completions

pakrym-oai committed:
share mode logic

pakrym-oai committed:
model_provider_info: fix auth arg usage and endpoint URL for ChatGPT mode

pakrym-oai committed:
test: fix auth header test and use safe env var for provider

- Relax Authorization header regex to accept any Bearer token.
- Use the existing 'USER' environment variable for provider API key to avoid unsafe code.
- Remove unsafe set_var for test API key.

pakrym-oai committed:
tests: update azure_overrides_assign_properties_used_for_responses_url to use PATH env variable instead of USER

bolinfest commented on codex-rs/core/src/model_provider_info.rs:
```diff
@@ -88,25 +90,31 @@ impl ModelProviderInfo {
+        auth: &Option<CodexAuth>,
     ) -> crate::error::Result<reqwest::RequestBuilder> {
-        let url = self.get_full_url();
+        let mut auth = auth;
+        let fallback_auth;
+        if auth.is_none() {
+            fallback_auth = self.get_fallback_auth()?;
+            auth = &fallback_auth;
+        }
```
This doesn't seem right. I asked chat about it and it suggested:

```rust
use std::borrow::Cow;

let auth: Cow<'_, Auth> = match auth {
    Some(a) => Cow::Borrowed(a),
    None => Cow::Owned(self.get_fallback_auth()?),
};

// Now you can use `auth` anywhere a `&Auth` is needed
do_something_with(&*auth);
```

bolinfest commented on codex-rs/core/src/model_provider_info.rs:
```diff
@@ -88,25 +90,31 @@ impl ModelProviderInfo {
-    pub(crate) fn get_full_url(&self) -> String {
-        let query_string = self
-            .query_params
+    pub fn get_query_string(&self) -> String {
```
Should this be private now?

bolinfest commented on codex-rs/core/tests/client.rs:
```diff
@@ -376,6 +379,72 @@ async fn includes_user_instructions_message_in_request() {
+        name: "custom".to_string(),
+        base_url: Some(format!("{}/openai", server.uri())),
+        // Reuse the existing environment variable to avoid using unsafe code
+        env_key: Some("USER".to_string()),
```
`USERNAME` on Windows?

bolinfest commented on codex-rs/core/tests/client.rs:
```diff
@@ -376,6 +379,72 @@ async fn includes_user_instructions_message_in_request() {
+        .and(path("/openai/responses"))
+        .and(query_param("api-version", "2025-04-01-preview"))
+        .and(header_regex("Custom-Header", "Value"))
+        .and(header_regex("Authorization", "Bearer .+"))
```
If you read `USER`/`USERNAME` you can make this an exact value?

bolinfest commented on codex-rs/core/src/model_provider_info.rs:
```diff
@@ -115,16 +123,26 @@ impl ModelProviderInfo {
-                format!("{base_url}/responses{query_string}")
-            }
+            WireApi::Responses => match auth {
+                Some(CodexAuth {
+                    mode: AuthMode::ChatGPT,
+                    ..
+                }) => {
+                    format!("{base_url}/responses{query_string}")
+                }
+                _ => format!("{base_url}/responses{query_string}"),
```
Both cases are the same here, no?

bolinfest reviewed (approved):
I would go with:

```rust
let user_env_var = if cfg!(windows) { "USERNAME" } else { "USER" };
```

just because `PATH` can be long and is maybe going to cause some dev somewhere a problem one day when running the unit tests.

I'll leave it up to you, though. I trust you to address the outstanding comments.

pakrym-oai committed:
client: simplify ChatGPT auth header assignment using if-let chain

model_provider_info: use Cow for auth fallback, make some methods private, clean up base_url logic

tests: use more robust env var for Azure overrides test
